engine/kv_engine.cpp

```diff
@@ -1324,6 +1327,14 @@ Status KVEngine::BatchWrite(const WriteBatch& write_batch) {
   std::set<SpinMutex*> spins_to_lock;
   std::vector<BatchWriteHint> batch_hints(write_batch.Size());
   std::vector<uint64_t> space_entry_offsets;
+  auto free_space_on_failure = [&]() {
+    for (size_t i = 0; i < batch_hints.size(); i++) {
+      if (batch_hints[i].allocated_space.size > 0) {
+        this->markEmptySpace(batch_hints[i].allocated_space);
```

Comment:
persistSpaceEntry()?

Reply:
merged in pmem allocator as a public function MarkFreeSpace